hoststats: add package for collecting host statistics including cpu memory and disk usage

nickethier opened the pull request:
### Description

This PR adds the ability for Consul to report host statistics through go-metrics. These metrics include cpu utilization by core and in aggregate, system memory usage and disk usage for the disk backing the configured data_dir. 

### Testing & Reproduction steps

Tests were added to assert cpu utilization calculation is performed accurately. Manual testing was performed on linux and darwin to ensure reported host stats values were accurate.

### PR Checklist

* [x] updated test coverage
* [x] external facing docs updated
* [x] appropriate backport labels added
* [x] not a security concern

loshz commented on .changelog/17038.txt:
````diff
@@ -0,0 +1,3 @@
+```release-note:improvement
+agent: add new metrics to track cpu disk and memory usage for server hosts
````
```suggestion
agent: add new metrics to track cpu disk and memory usage for server hosts (defaults to: X)
```

loshz commented on agent/setup.go:
```diff
@@ -4,13 +4,15 @@
 	"time"
 
 	"github.com/armon/go-metrics/prometheus"
+	"github.com/hashicorp/consul/lib/hoststats"
```
Nit: needs to be grouped with the other hashicorp/consul imports.

loshz commented on agent/setup.go:
```diff
@@ -58,6 +60,7 @@ type BaseDeps struct {
 	WatchedFiles  []string
 
 	deregisterBalancer, deregisterResolver func()
+	stopHostCollector                      func()
```
```suggestion
	stopHostCollector                      context.CancelFunc
```

loshz commented on lib/hoststats/collector.go:
```diff
@@ -0,0 +1,189 @@
+	return collector
+}
+
+func (h *Collector) loop(ctx context.Context) {
```
Nit:
```suggestion
func (c *Collector) loop(ctx context.Context) {
```

loshz commented on lib/hoststats/collector.go:
```diff
@@ -0,0 +1,189 @@
+
+// collect will collect stats related to resource usage of the host
+func (h *Collector) collect() {
+	h.hostStatsLock.Lock()
+	defer h.hostStatsLock.Unlock()
```
You can probably move this to the end of this method so you don't hold the lock while you collect metrics:
```go
	// Update the collected status object.
	h.hostStatsLock.Lock()
	h.hostStats = hs
	h.hostStatsLock.Unlock()

loshz commented on lib/hoststats/cpu.go:
```diff
@@ -0,0 +1,83 @@
+// calculate calculates the current cpu usage percentages
+func (h *cpuStatsCalculator) calculate(times cpu.TimesStat) *CPUStats {
+
+	currentBusy := times.User + times.System + times.Nice + times.Iowait + times.Irq +
+		times.Softirq + times.Steal + times.Guest + times.GuestNice
+	currentTotal := currentBusy + times.Idle
```
Need to add a comment around this whole logic and how we got here.

loshz commented on lib/hoststats/cpu.go:
```diff
@@ -0,0 +1,83 @@
+		times.Softirq + times.Steal + times.Guest + times.GuestNice
+	currentTotal := currentBusy + times.Idle
+
+	deltaTotal := currentTotal - h.prevTotal
+	stats := &CPUStats{
+		CPU: times.CPU,
+
+		Idle:   ((times.Idle - h.prev.Idle) / deltaTotal) * 100,
+		User:   ((times.User - h.prev.User) / deltaTotal) * 100,
+		System: ((times.System - h.prev.System) / deltaTotal) * 100,
+		Iowait: ((times.Iowait - h.prev.Iowait) / deltaTotal) * 100,
+		Total:  ((currentBusy - h.prevBusy) / deltaTotal) * 100,
+	}
```
Same here, needs a comment.

loshz commented on lib/hoststats/cpu_test.go:
```diff
@@ -0,0 +1,60 @@
+	"github.com/stretchr/testify/require"
+)
+
+func TestHostStats_CPU(t *testing.T) {
```
I'm not sure of a better approach, but I don't like the thought of this being a unit test. We're essentially spinning a CPU core to generate some load, and this took almost 13s to run on my local machine.

Maybe we need a whole new integration test suite for this kind of stuff?

loshz commented on lib/hoststats/host.go:
```diff
@@ -0,0 +1,97 @@
+	Timestamp    int64
+}
+
+func (hs *HostStats) Clone() *HostStats {
+	clone := *hs
+
+	clone.CPU = make([]*CPUStats, len(hs.CPU))
+	for i := range hs.CPU {
+		cpu := *hs.CPU[i]
+		clone.CPU[i] = &cpu
+	}
+	return &clone
+}
```
I think you can just do this:
```suggestion
func (hs *HostStats) Clone() *HostStats {
	clone := &HostStats{}
	*clone = *hs
	return clone
}
```

https://go.dev/play/p/1Uiyq3Dcq4m

loshz commented on website/content/docs/agent/telemetry.mdx:
```diff
@@ -755,3 +755,28 @@ Consul attaches the following labels to metric values.
 
+## Server Host Metrics
+
+Consul servers report the following metrics about the host's system resources
```
We should add a note about how this currently only works on VMs, so there are no surprises to users if they attempt to collect metrics inside a container.

loshz reviewed (changes requested):
Overall, I think this is a good first attempt that we can iterate on in the future. I've left a couple of nits/change requests, and I think you might have missed a few steps from [this](https://github.com/hashicorp/consul/blob/main/docs/config/checklist-adding-config-fields.md) doc.

Couple of open questions:
- Are we defaulting to host metrics being enabled or disabled?
- Can you upload some sort of sample of the metrics? As right now it's hard for me to test this locally.
- We should really add some sort of integration testing strategy. I know it's borderline impossible to assert these metrics, but we can at least attempt to run the collector on different platforms.
- We should probably add a document that outlines the metrics we are collecting and the format as well as reasoning.

nickethier commented:
Thanks for the review @loshz! I addressed your comments and answered your questions below.

>Overall, I think this is a good first attempt that we can iterate on in the future. I've left a couple of nits/change requests, >and I think you might have missed a few steps from [this](https://github.com/hashicorp/consul/blob/main/docs/config/checklist-adding-config-fields.md) doc.
>

Hmm I'm not sure I see a step I missed. There is no `RuntimeConfig` change because the field gets parsed into the `TelemetryConfig` struct like other telemetry fields.

> Couple of open questions:
> 
> * Are we defaulting to host metrics being enabled or disabled?

Yes defaulting to enabled. `go-psutil` takes care of the platform specifics.

> * Can you upload some sort of sample of the metrics? As right now it's hard for me to test this locally.

Sure heres what I've been testing with:
```
# starting a dev agent some extra config to enable prometheus format
$> consul agent -dev -hcl 'telemetry{ prometheus_retention_time = "1h" disable_hostname = true } data_dir="/Users/nethier/tmp"'

# in another terminal, query the metrics for any consul_host prefixed metrics
$> curl 'http://localhost:8500/v1/agent/metrics?format=prometheus' | grep consul_host
# HELP consul_host_cpu_idle Idle cpu utilization
# TYPE consul_host_cpu_idle gauge
consul_host_cpu_idle 0
consul_host_cpu_idle{cpu="cpu0"} 81.93743896484375
consul_host_cpu_idle{cpu="cpu1"} 99.5999984741211
consul_host_cpu_idle{cpu="cpu10"} 94.08817291259766
consul_host_cpu_idle{cpu="cpu11"} 99.69999694824219
consul_host_cpu_idle{cpu="cpu12"} 93.98797607421875
consul_host_cpu_idle{cpu="cpu13"} 99.69999694824219
consul_host_cpu_idle{cpu="cpu14"} 96.39639282226562
consul_host_cpu_idle{cpu="cpu15"} 99.80000305175781
consul_host_cpu_idle{cpu="cpu2"} 86.8951644897461
consul_host_cpu_idle{cpu="cpu3"} 99.69969940185547
consul_host_cpu_idle{cpu="cpu4"} 89.74874114990234
consul_host_cpu_idle{cpu="cpu5"} 99.79959869384766
consul_host_cpu_idle{cpu="cpu6"} 91.3741226196289
consul_host_cpu_idle{cpu="cpu7"} 99.69999694824219
consul_host_cpu_idle{cpu="cpu8"} 92.8786392211914
consul_host_cpu_idle{cpu="cpu9"} 99.60040283203125
# HELP consul_host_cpu_iowait Iowait cpu utilization
# TYPE consul_host_cpu_iowait gauge
consul_host_cpu_iowait 0
consul_host_cpu_iowait{cpu="cpu0"} 0
consul_host_cpu_iowait{cpu="cpu1"} 0
consul_host_cpu_iowait{cpu="cpu10"} 0
consul_host_cpu_iowait{cpu="cpu11"} 0
consul_host_cpu_iowait{cpu="cpu12"} 0
consul_host_cpu_iowait{cpu="cpu13"} 0
consul_host_cpu_iowait{cpu="cpu14"} 0
consul_host_cpu_iowait{cpu="cpu15"} 0
consul_host_cpu_iowait{cpu="cpu2"} 0
consul_host_cpu_iowait{cpu="cpu3"} 0
consul_host_cpu_iowait{cpu="cpu4"} 0
consul_host_cpu_iowait{cpu="cpu5"} 0
consul_host_cpu_iowait{cpu="cpu6"} 0
consul_host_cpu_iowait{cpu="cpu7"} 0
consul_host_cpu_iowait{cpu="cpu8"} 0
consul_host_cpu_iowait{cpu="cpu9"} 0
# HELP consul_host_cpu_system System cpu utilization
# TYPE consul_host_cpu_system gauge
consul_host_cpu_system 0
consul_host_cpu_system{cpu="cpu0"} 6.35721492767334
consul_host_cpu_system{cpu="cpu1"} 0.30000001192092896
consul_host_cpu_system{cpu="cpu10"} 2.605210304260254
consul_host_cpu_system{cpu="cpu11"} 0.20000000298023224
consul_host_cpu_system{cpu="cpu12"} 2.8056111335754395
consul_host_cpu_system{cpu="cpu13"} 0.30000001192092896
consul_host_cpu_system{cpu="cpu14"} 1.30130136013031
consul_host_cpu_system{cpu="cpu15"} 0.20000000298023224
consul_host_cpu_system{cpu="cpu2"} 4.536290168762207
consul_host_cpu_system{cpu="cpu3"} 0.20020020008087158
consul_host_cpu_system{cpu="cpu4"} 3.718592882156372
consul_host_cpu_system{cpu="cpu5"} 0.20040079951286316
consul_host_cpu_system{cpu="cpu6"} 2.808425188064575
consul_host_cpu_system{cpu="cpu7"} 0.20000000298023224
consul_host_cpu_system{cpu="cpu8"} 2.407221555709839
consul_host_cpu_system{cpu="cpu9"} 0.2997002899646759
# HELP consul_host_cpu_total Total cpu utilization
# TYPE consul_host_cpu_total gauge
consul_host_cpu_total 0
consul_host_cpu_total{cpu="cpu0"} 18.062562942504883
consul_host_cpu_total{cpu="cpu1"} 0.4000000059604645
consul_host_cpu_total{cpu="cpu10"} 5.911823749542236
consul_host_cpu_total{cpu="cpu11"} 0.30000001192092896
consul_host_cpu_total{cpu="cpu12"} 6.01202392578125
consul_host_cpu_total{cpu="cpu13"} 0.30000001192092896
consul_host_cpu_total{cpu="cpu14"} 3.6036036014556885
consul_host_cpu_total{cpu="cpu15"} 0.20000000298023224
consul_host_cpu_total{cpu="cpu2"} 13.104838371276855
consul_host_cpu_total{cpu="cpu3"} 0.3003003001213074
consul_host_cpu_total{cpu="cpu4"} 10.251255989074707
consul_host_cpu_total{cpu="cpu5"} 0.20040079951286316
consul_host_cpu_total{cpu="cpu6"} 8.625877380371094
consul_host_cpu_total{cpu="cpu7"} 0.30000001192092896
consul_host_cpu_total{cpu="cpu8"} 7.121364116668701
consul_host_cpu_total{cpu="cpu9"} 0.39960038661956787
# HELP consul_host_cpu_user User cpu utilization
# TYPE consul_host_cpu_user gauge
consul_host_cpu_user 0
consul_host_cpu_user{cpu="cpu0"} 11.705348014831543
consul_host_cpu_user{cpu="cpu1"} 0.10000000149011612
consul_host_cpu_user{cpu="cpu10"} 3.3066132068634033
consul_host_cpu_user{cpu="cpu11"} 0.10000000149011612
consul_host_cpu_user{cpu="cpu12"} 3.2064127922058105
consul_host_cpu_user{cpu="cpu13"} 0
consul_host_cpu_user{cpu="cpu14"} 2.302302360534668
consul_host_cpu_user{cpu="cpu15"} 0
consul_host_cpu_user{cpu="cpu2"} 8.568548202514648
consul_host_cpu_user{cpu="cpu3"} 0.10010010004043579
consul_host_cpu_user{cpu="cpu4"} 6.532663345336914
consul_host_cpu_user{cpu="cpu5"} 0
consul_host_cpu_user{cpu="cpu6"} 5.817452430725098
consul_host_cpu_user{cpu="cpu7"} 0.10000000149011612
consul_host_cpu_user{cpu="cpu8"} 4.714142322540283
consul_host_cpu_user{cpu="cpu9"} 0.09990009665489197
# HELP consul_host_disk_available Available bytes on disk
# TYPE consul_host_disk_available gauge
consul_host_disk_available 0
consul_host_disk_available{path="/Users/nethier/tmp"} 7.7038125056e+10
# HELP consul_host_disk_inodes_percent Percentage of disk inodes usage
# TYPE consul_host_disk_inodes_percent gauge
consul_host_disk_inodes_percent 0
consul_host_disk_inodes_percent{path="/Users/nethier/tmp"} 0.5227161049842834
# HELP consul_host_disk_size Size of disk in bytes
# TYPE consul_host_disk_size gauge
consul_host_disk_size 0
consul_host_disk_size{path="/Users/nethier/tmp"} 4.9996316672e+11
# HELP consul_host_disk_used Disk usage in bytes
# TYPE consul_host_disk_used gauge
consul_host_disk_used 0
consul_host_disk_used{path="/Users/nethier/tmp"} 4.22925041664e+11
# HELP consul_host_disk_used_percent Percentage of disk space usage
# TYPE consul_host_disk_used_percent gauge
consul_host_disk_used_percent 0
consul_host_disk_used_percent{path="/Users/nethier/tmp"} 84.59123992919922
# HELP consul_host_memory_available Available physical memory in bytes
# TYPE consul_host_memory_available gauge
consul_host_memory_available 1.3143601152e+10
# HELP consul_host_memory_free Free physical memory in bytes
# TYPE consul_host_memory_free gauge
consul_host_memory_free 3.174227968e+09
# HELP consul_host_memory_total Total physical memory in bytes
# TYPE consul_host_memory_total gauge
consul_host_memory_total 3.4359738368e+10
# HELP consul_host_memory_used Used physical memory in bytes
# TYPE consul_host_memory_used gauge
consul_host_memory_used 2.1216137216e+10
# HELP consul_host_memory_used_percent Percentage of physical memory in use
# TYPE consul_host_memory_used_percent gauge
consul_host_memory_used_percent 61.74708557128906
# HELP consul_host_uptime System uptime
# TYPE consul_host_uptime gauge
consul_host_uptime 633771
```

> * We should really add some sort of integration testing strategy. I know it's borderline impossible to assert these metrics, but we can at least attempt to run the collector on different platforms.

I refactored the test a bit to be more conservative. I'm hoping long term we can cover this in our e2e testing of the hcp telemetry components.

> * We should probably add a document that outlines the metrics we are collecting and the format as well as reasoning.

Any suggestions or prior art to help give me an idea of where and what level of detail something like this would be?

dhiaayachi commented:
👋 @nickethier,
This LGTM code wise. 

Can you please give more details on the reasoning behind adding those metrics to be collected by Consul versus using another mean of collecting the host data (an external exporter/collector)?

The reason behind my question is that I think this add an extra concern around architecture compatibility. I see that the used library have a wide compatibility architecture range, but from Consul perspective we still need to verify that all the architectures that we support are working correctly and currently we don't have a great coverage for those different architectures. 

I'm also not sure how accurate those metrics will be in K8S or in containers in general 🤔

nickethier commented:
> The reason behind my question is that I think this add an extra concern around architecture compatibility. I see that the used library have a wide compatibility architecture range, but from Consul perspective we still need to verify that all the architectures that we support are working correctly and currently we don't have a great coverage for those different architectures.

As I understand this library (`gopsutil`) has been used in Consul for many years. However we are likely calling functions that had not been used before. That being said it is also used across many different architectures for similar purposes in Nomad, where [the origins of this code was inspired from](https://github.com/hashicorp/nomad/blob/main/client/stats/cpu.go).

In cases where there isn't platform support I expect the metric to read as `0`. For example the above sample is from my local machine running macOS. On this platform cpu `iowait` time is not supported. 

> 
> I'm also not sure how accurate those metrics will be in K8S or in containers in general 🤔

I called this out in the docs. Perhaps there is room in the future to iterate on this and report any container limits in place but this is very platform specific and not something `gopsutil` supports and would have to implemented.

dhiaayachi reviewed (approved):
Thank you for answering my questions @nickethier 🙏 

LGTM!
